---
title: "Glycosyltransferases and glycosidases in Treg activation"
author: "Marek Gierlinski"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    css: tabbed_notebook.css
  pdf_document:
    number_sections: yes
---


```{r options, echo=FALSE}
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  echo = FALSE,
  cache = TRUE
)
options(width = 100)
options(knitr.table.format = "html")
```

```{css echo=FALSE}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

```{r kable_functions}
myKable <- function(df, row.names=FALSE, col.names=NA, digits=2, bootstrap="condensed", font.size=12) {
  kable(df, format="html", row.names=row.names, col.names=col.names, digits=digits) %>% kable_styling(bootstrap_options=bootstrap, full_width=FALSE, position="left", font_size=font.size)
}
```

Collaborators: [Paul Crocker](), [Gavuthami Murugesan](), [Gang Wu]()

[Proposal](http://www.compbio.dundee.ac.uk/user/mgierlinski/glycotreg/doc/Proposal.pdf)

[Sequencing report](http://www.compbio.dundee.ac.uk/user/mgierlinski/glycotreg/doc/E181945_report.pdf)

Notebooks: [1](http://www.compbio.dundee.ac.uk/user/mgierlinski/glycotreg/doc/Notes.1.nb.html)


[Interactive data explorer](https://shiny.compbio.dundee.ac.uk/private/marek_glycotreg/shiny_de/)

```{r global_setup, cache=FALSE, echo=FALSE}
source("../R/setup.R")
source("../R/star.R")
source("../R/plots.R")
source("../R/biomart.R")
source("../R/set_enrichment.R")
library(tidyverse)
library(knitr)
library(kableExtra)
library(ggbeeswarm)
library(GGally)
library(cowplot)
library(readxl)
#library(grid)
library(viridis)
library(biomaRt)
library(edgeR)
```

# {.tabset}


## Data

### Data summary

The samples have been sequenced by Edinburgh Clinical Research Facility on NextGen 550 platform. There are 8 samples in total. Reads are paired-end 2 x 75 bp. The FASTQ files were made available on Illumina's BaseSpace. 

```{r prepare_metadata}
metadata <- tibble(
  condition = c(rep("TregRest", 4), rep("TregAct", 4)),
  replicate = c(1:4, 1:4)
) %>% 
  mutate(sample = paste0(condition, "_", replicate)) %>%
  mutate(name = sample, group = factor(condition, levels=c("TregRest", "TregAct"))) %>%
  as.data.frame
```

The experimental design is very simple: two conditions in four replicates each. The conditions are regulatory T cells (Tregs) at rest and activated.

```{r metadata_show}
myKable(metadata[, c("condition", "replicate", "sample")]) 
```

### Data processing

Basic data processing is done by a [snakemake script](http://www.compbio.dundee.ac.uk/user/mgierlinski/abin1/Snakefile) in a `conda` environment. These steps can be reproduced by:

```
conda create --name glycotreg --file conda-spec.txt
source activate glycotreg
snakemake -c "qsub -V -cwd -o snakelog -e snakelog -pe smp {threads}" --jobs=100
```
The `snakemake` processing consists of the following steps:

- load genome from Ensembl
- merge FASTQ lanes
- do *fastqc* and *multiqc* quality control
- create *STAR* index
- run *STAR* aligner and quantifier
- create bedGraph files for overview of mapping

### Quality control

We use [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) and [MultiQC](http://multiqc.info) for quality control. The full QC report is available [here](`r paste0(public_html, "/multiqc/report.html")`).

Briefly, quality of data is generally good. We have 50-60 M reads per sample, read quality is very good. 

## Mapping

### Genome, transcriptome and gene annotations

We use mouse genome data from Ensembl, assembly GRCm38, release 96.

```{r genes_biomart}
# there is a curl error when accessing specific version 96. Perhaps a temporary glitch.
# Instead, I ignore version number and use the current version. At the moment of writing this script 
# it is 96.
#mart <- useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl", version=96)

mart <- useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl", version = 96)
 
genes <- biomartGeneDownload(mart)
genes <- genes[, c("gene_id", "gene_name", "description")]
genes$description <- gsub("\\s\\[.*\\]", "", genes$description, perl=TRUE)

#transcripts <- biomartTranscriptDownload(mart)
#transcripts <- transcripts[, c("transcript_id", "transcript_name", "gene_name", "description")]
#transcripts$description <- gsub("\\s\\[.*\\]", "", transcripts$description, perl=TRUE)
```

For genome sequences, we downloaded RepeatMasker-filtered primary assembly. The gene models are in the form of a GTF file. 

#### Glycogenes

I received a table with glycogenes from Gavuthami.

```{r read_glycogenes, eval=FALSE}
# old file in different format
file <- paste0("../data/glycogenes.txt")
glyco <- read.table(file, sep="\t", header=TRUE)

# there are weird strings in column names
colnames(glyco) <- gsub("\\.M20.*$", "", colnames(glyco), perl=TRUE)

# some refseq entries have two comma-separated IDs
refseq <- glyco$yourlist %>% as.character %>% unique %>% strsplit(",") %>% unlist
```

```{r convert_glyco, eval=FALSE}
# old
glyco.map <- getBM(
  attributes=c("refseq_mrna", "ensembl_gene_id", "external_gene_name"),
  filters = "refseq_mrna",
  values = refseq,
  mart = mart
)
glyco.ens <- unique(glyco.map$ensembl_gene_id)
```

```{r read_glycogenes_new}
glyco.ens <- read_excel("../data/List of glycogenes and functions.xlsx") %>% 
  pull(ensembl_id)
```


```{r glyco_ens}
myKable(tibble(ensembl_id = glyco.ens)) %>% scroll_box(height = "400px", width = "400px")
```

This gives us `r N(length(glyco.ens))` unique Ensembl IDs associated with glycogenes.

### STAR

We mapped reads to the genome using STAR, we also calculated reads per gene during mapping step.

#### Mapping properties

The table below summarises mapping properties. The percentage of uniquely mapped reads is around 80%, which is usual for mouse/human genome.

```{r star_properties}
metadata$sample %>%
  lapply(function(s) parseStarLog(dirs$starmap, s)) %>%
  do.call(cbind, .) %>%
  t %>%
  as.data.frame %>%
  dplyr::select(5, 7, 8, 9, 10, 11, 21:27) %>%
  kable() %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), font_size = 10) %>%
  scroll_box(width = "900px")
```

#### Strandedness

Next, we checked if data were stranded as they should be. We read STAR gene read count file for one sample (`TregRest_1`). It contains three columns of counts: 

- unstranded RNA-seq
- 1st read strand aligned with RNA
- 2nd read strand aligned with RNA

Depending on the protocol, column 2 (or 3) contains sense reads, while the other column (3 or 2) should contain antisense reads. We expect the antisense reads to be much less than sense reads.

```{r star_strandedness, fig.width=8, fig.height=2.5}
plotStarSense(file.path(dirs$readcount, "TregRest_1.txt"))
```

The second read strand contains our stranded data.

#### Read distribution

```{r read_bedgraphs_ens}
bg <- readBedgraphs(dirs$bedgraph, metadata, windowSize = 1000000)
```

Here we show reads mapped to the entire genome binned in bins of 1 Mb. The horizontal axis shows the cumulative genomic position across all chromosomes. Colour (not very well visible) show alternating chromosomes. This is a sanity check, to make sure there is nothing weird with data (e.g. excessive ribosomal sequences).

```{r manhattan_plot_ens, fig.width=10, fig.height=8}
plotManhattan(bg, yscale=1e6) 
```

It looks nice and clean. The strong peak on the right is the start of MT, so probably not interesting. The peak on the left is in chromosome 1, bin 24.0-24.5 Mb. We might check it later to see what gene it is.

Here is the same plot, but with the vertical axis limited to show distribution of lower counts.

```{r manhattan_plot_ens_lim, fig.width=10, fig.height=8}
plotManhattan(bg, yscale=1e6, ymax=1) 
```

Nothing to complain about.

### Read count per gene

Let's read STAR read counts per gene and have a global look at all data.

```{r star_read_counts}
cnt.star <- do.call(cbind, lapply(metadata$sample, function(s) parseStarCount(dirs$readcount, s, column=4)))
cnt.star <- as.matrix(cnt.star)
colnames(cnt.star) <- metadata$name
tot.count <- colSums(cnt.star)
norm.fac <- tot.count / mean(tot.count)
cnt.star.norm <- as.data.frame(t(t(cnt.star) / norm.fac))
mean.star <- funCondition(cnt.star.norm, metadata, mean)
sd.star <- funCondition(cnt.star.norm, metadata, sd)
```

#### Distance matrix

The distance matrix show the Pearson's correlation coefficient between each pair of replicates, based on the read count per gene. 

```{r star_distance_matrix, fig.width=6, fig.height=5}
plotDistanceMatrix(cnt.star, metadata) 
```

At first glance, we have good data with reproducible replicates and big changes between biological conditions.

#### Clustering

This is confirmed by clustering.

```{r star_clustering, fig.width=5, fig.height=4}
plotClustering(cnt.star)
```

Note that there seem to be differences between replicates 1-2 and 3-4 in both conditions. Sequencing was done in one go, on a single flow cell. Perhaps this is due to how samples were prepared? I'm guessing replicates 1 and 2 (for both conditions) were prepared on one day and 3 and 4 on another. Is this right?

#### PCA

```{r star_pca, fig.width=6, fig.height=4}
plotPCA(cnt.star.norm, metadata)
```

#### Replicate comparison

```{r star_ens_xy_plot, fig.width=8, fig.height=7}
X <- cnt.star.norm
X[X == 0] <- NA
ggpairs(log10(X),
  lower = list(continuous=wrap("points", size=0.1, alpha=0.3)),
  upper = list(continuous=wrap("cor", size=3))
)
```


#### Read counts table

```{r star_read_counts_table}
cnt.genes <- merge(round(cnt.star.norm, 1), genes, by.x="row.names", by.y="gene_id")
colnames(cnt.genes)[1] <- "ensembl_id"

fname <- "star_counts_norm.txt"
if(!dir.exists(dirs$download)) dir.create(dirs$download)
file <- paste0(dirs$download, fname)
write.table(cnt.genes, file=file, quote=FALSE, sep="\t", col.names=TRUE, row.names=FALSE)

url <- paste0(public_html, "download/", fname)
```

The table with normalized counts per gene can be downloaded from [here](`r url`).

## Differential expression

We use `edgeR` with a simple two-condition design.

```{r design_setup}
design <- model.matrix(~0 + metadata$group)
colnames(design) <- levels(metadata$group)

gt.contrasts <- makeContrasts(
  "Active-Rest" = "TregAct-TregRest",
  levels = design
)
contrasts <- colnames(gt.contrasts)
contrast.names <- setNames(list(contrasts), contrasts)  # for use in figures; change in case contrasts are complicated
```


```{r edger_star_fit}
y <- DGEList(counts=as.matrix(cnt.star), group=metadata$group)
y <- calcNormFactors(y)
y <- estimateDisp(y, design)
star.fit <- glmQLFit(y, design) 
```

```{r edger_star_test}
star.test <- lapply(contrasts, function(ctr) {
  qlf <- glmQLFTest(star.fit, contrast=gt.contrasts[, ctr])
  tab <- topTags(qlf, n=1e16, adjust.method = "BH", sort.by = "none")$table
  tab <- merge(tab, genes, by.x="row.names", by.y="gene_id")
  colnames(tab)[1] <- "gene_id"
  tab
})
names(star.test) <- contrasts 
```


```{r star_volcano_plot_function}
plotVolcano <- function(tab, alpha=0.05, title="", sel=NULL) {
  lim <- log10(max(tab[tab$FDR < alpha, "PValue"]))
  tab$x <- tab$logFC
  tab$y <- -log10(tab$PValue)
  if(!is.null(sel)) {tab$sel <- tab$gene_id %in% sel} else {tab$sel=FALSE}
  ggplot(tab) +
    theme_classic() +
    theme(legend.position = "none") +
    geom_point(aes(x=x, y=y), colour="grey70", size=0.3) +
    geom_point(data=tab[tab$sel, ], aes(x=x, y=y), colour="black", size=0.6) +
    geom_vline(xintercept = 0, colour="grey70") +
    geom_hline(yintercept = -lim, colour="red", linetype="dashed") +
    labs(x=expression(log[2]~FC), y=expression(log[10]~P), title=title) +
    scale_y_continuous(expand=c(0,0), limits=c(0, 1.04*max(-log10(tab$PValue))))
}

plotPDist <- function(tab, title="", bin.size=0.01) {
  ggplot(tab) +
    theme_classic() +
    geom_histogram(aes(x=PValue, y=..density..), breaks=seq(0,1,bin.size), fill="darkgoldenrod2") +
    labs(x="P-value", y="Density", title=title)
}
```

### P-value distribution

```{r p_distribution, fig.width=5, fig.height=5}
x <- star.test[[1]] %>% filter(logCPM > 0)
plotPDist(x, bin.size=0.005)
```


### Volcano plot

Grey points show all data, black points are glycogenes. In-depth analysis of this plot can be done using [interactive data explorer](https://shiny.compbio.dundee.ac.uk/marek_glycotreg/shiny_de/).

```{r star_volcano_plots, fig.width=6, fig.height=5}
P <- lapply(names(star.test), function(ctr) {
  plotVolcano(star.test[[ctr]], title=ctr, sel=glyco.ens)
})
plot_grid(plotlist = P, nrow = 1)
 
p.val <- star.test[[1]]$FDR 
```

There are `r N(sum(p.val < 0.05))` differentially expressed genes (at FDR level of 0.05) among `r N(length(p.val))` of all tested genes.

### DE tables

The link below provides a tab-delimited text table with differential expression results for all genes.

```{r star_DE_tables}
links <- data.frame()
if(!dir.exists(dirs$download)) dir.create(dirs$download)
for(ctr in contrasts) {
  fname <- paste0("star_de_", ctr, ".txt")
  file <- paste0(dirs$download, fname)
  write.table(star.test[[contrast.names[[ctr]]]], file=file, quote=FALSE, sep="\t", col.names=TRUE, row.names=FALSE)
  links <- rbind(links, data.frame(
    url = paste0("[", contrast.names[[ctr]], "](", public_html, "download/", fname, ")")  
  ))
} 
myKable(links)
```



### Functional enrichment

We download GO-term data from `BioMart`, using the same version (96) as for our genome. For pathway analysis we download data from Reactome, using the current file: https://reactome.org/download/current/Ensembl2Reactome.txt. Both data sets are uploaded to the Shiny server and are used in the [interactive data explorer](https://shiny.compbio.dundee.ac.uk/marek_glycotreg/shiny_de/).

```{r go_term_names}
go.terms <- biomartGODescriptions(mart)
```

```{r gene_go_terms}
gene_ids <- row.names(cnt.star)
gene.go <- biomartGODownload(mart, gene_ids)
```

```{r go_set_enrichment}
gene.go.enr <- setNames(gene.go, c("id", "term"))
go.terms.enr <- setNames(go.terms, c("term", "name", "domain"))
P <- parallel::mclapply(star.test, function(tab) {
  scores <- setNames(abs(tab$logFC), tab$gene_id)
  star.go.set <- setEnrichment(scores, gene.go.enr, go.terms.enr, nboot=1000)
}, mc.cores = 3)
names(P) <- names(star.test)
```

```{r read_reactome}
# reactome <- read.delim("https://reactome.org/download/current/Ensembl2Reactome.txt", quote="", header=FALSE, sep="\t", stringsAsFactors=FALSE)
# names(reactome) <- c("gene_id", "reactome_id", "URL", "name", "evidence", "species")
reactome <- read_tsv("../cache/reactome.tsv")
```

```{r filter_reactome}
gene_ids <- row.names(cnt.star)
reactome.sel <- reactome[reactome$gene_id %in% gene_ids,]
```

```{r process_reactome}
reactome.terms <- unique(reactome[, c("reactome_id", "name")])
gene.reactome <- reactome.sel[, c("gene_id", "reactome_id")]
```


```{r save_for_shiny_de}
if(!dir.exists("../RData")) dir.create("../RData")
DE <- star.test
expr <- cnt.star.norm
metadata$condition <- factor(metadata$condition)
metadata$condition <- relevel(metadata$condition, ref="TregRest")
# need version 2 to be compatible with R 3.3 on the Shiny server
save(DE, expr, go.terms, gene.go, reactome.terms, gene.reactome, metadata, glyco.ens, file="../RData/shiny_data.RData", version=2)
rm(DE, expr)
```


### Gene set enrichment

Set enrichment using R package `fgsea`.

```{r gsea}
library(fgsea)
select <- dplyr::select

go_terms <- list(
  gene2term = gene.go %>% rename(term_id = go_id),
  terms = go.terms %>% rename(term_id = go_id, term_name = go_name)
)
re_terms <- list(
  gene2term = gene.reactome %>% rename(term_id = reactome_id),
  terms = reactome.terms %>% rename(term_id = reactome_id, term_name = name)
)
de <- star.test$`Active-Rest` %>% as_tibble() %>% filter(logFC != 0)

fg_go <- fgsea_cache(de, go_terms, file = "../cache/fg_go.rds")
fg_re <- fgsea_cache(de, re_terms, file = "../cache/fg_re.rds")
```

Terms and pathways related to glycogenes

```{r fgsea_glyco}
fg_go_sig <- fg_go %>% filter(padj < 0.05)
fg_re_sig <- fg_re %>% filter(padj < 0.05)

fg_go_gly <- fg_go_sig %>% filter(str_detect(term_name, "[gG]lyc"))
fg_re_gly <- fg_re_sig %>% filter(str_detect(term_name, "[gG]lyc"))

fg_go_sig %>% select(-leading_edge) %>% myKable()
fg_re_sig %>% select(-leading_edge) %>% myKable()
```

There are `r nrow(fg_go_sig)` significant GO terms and `r nrow(fg_re_sig)` significant reactome pathways. None of them are related to glycogenes.

## Glycogenes

```{r plot_one_gene_function}
plotOneGene <- function(dat, metadata, name="") {
  d <- data.frame(
    metadata,
    expr = as.numeric(dat)
  )
  
  ggplot(d, aes(x=condition, y=expr, fill=as.factor(replicate))) +
    theme_classic() +
    theme(
      legend.position = "none",
      plot.title = element_text(size=12)
    ) +
    geom_beeswarm(cex=5, size=3, shape=21) +
    scale_y_continuous(expand=c(0,0), limits=c(0, 1.05*max(d$expr))) +
    #scale_fill_manual(values=cbPalette) +
    viridis::scale_fill_viridis(discrete=TRUE, option="cividis") +
    labs(x=NULL, y=NULL, title=name)
}
```

```{r glycosel}
DE <- star.test$`Active-Rest`
rownames(DE) <- DE$gene_id
sel <- DE$FDR < 0.05 & DE$gene_id %in% glyco.ens
glyco.sel <- DE[sel, "gene_id"]
glyco.sel <- glyco.sel[order(DE[glyco.sel, "logFC"])]
```

Out of the `r length(glyco.ens)` glycogenes we have `r length(glyco.sel)` genes that are differentially expressed. Here I show intensity plots for all of them. The number next to the gene name is log-2 fold change.

```{r plot_glycogenes, fig.width=12, fig.height=2.5}
lst <- split(glyco.sel, (0:length(glyco.sel) %/% 6))
#lst <- lst[1:3]
F <- lapply(lst, function(genes) {
  P <- lapply(genes, function(gene) {
    d <- DE[DE$gene_id == gene,]
    s <- sprintf("%s  %5.2f", d$gene_name, d$logFC)
    plotOneGene(cnt.star.norm[gene, ], metadata, s)
  })
  if(length(P) > 0) {
    g <- plot_grid(plotlist = P, nrow = 1)
    print(g)
  }
})
```

## Session info

```{r session_info}
options(width = 80)
sessionInfo()
```
